Putting HashMap in a heap block hurts memory use, speed, code size (WebCore DOM)

darinadler opened the pull request:
#### 7945fc61f7720ebd0fa9a2bd7d047b6f0d0c2586
<pre>
Putting HashMap in a heap block hurts memory use, speed, code size (WebCore DOM)
<a href="https://rdar.apple.com/155739380">rdar://155739380</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=295864">https://bugs.webkit.org/show_bug.cgi?id=295864</a>

Reviewed by Chris Dumez.

HashMap and similar classes are already pointers to a hash table, so there is no reason
to wrap them in a unique_ptr in a heap block and add another level of indirection.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::buildAccessKeyCache): Updated since m_accessKeyCache is
std::optional instead of std::unique_ptr.
(WebCore::Document::invalidateAccessKeyCacheSlowCase): Ditto.
(WebCore::Document::willBeRemovedFromFrame): Updated since m_touchEventTargets
and m_wheelEventTargets are not stored in std::unique_ptr any more.
(WebCore::Document::absoluteRegionForWheelEventTargets): Renamed from
absoluteRegionForEventTargets and updated to use m_wheelEventTargets.

* Source/WebCore/dom/Document.h: Updated m_accessKeyCache to use
std::optional and m_touchEventTargets and m_wheelEventTargets to not
use std::unique_ptr. Deleted unused touchEventTargets and wheelEventTargets
functions. Made absoluteEventRegionForNode private and renamed
absoluteRegionForEventTargets to absoluteRegionForWheelEventTargets
since that&apos;s the only thing it&apos;s used for.

* Source/WebCore/page/DebugPageOverlays.cpp:
(WebCore::MouseWheelRegionOverlay::updateRegion): Use
absoluteRegionForWheelEventTargets.
* Source/WebCore/page/scrolling/ScrollingCoordinator.cpp:
(WebCore::ScrollingCoordinator::absoluteEventTrackingRegionsForFrame const):
Ditto.

Canonical link: <a href="https://commits.webkit.org/297735@main">https://commits.webkit.org/297735@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/84b2e2dd1495964661656232c647842c11b7d8b0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112618 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32350 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22828 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118817 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63135 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114580 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33002 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40913 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85710 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/36364 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115565 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26338 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101308 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66015 "Found 139 new API test failures: /WPE/TestLoaderClient:/webkit/WebKitWebView/loading-status, /WPE/TestWebKitWebView:/webkit/WebKitWebView/notification, /WPE/TestLoaderClient:/webkit/WebKitWebView/reload, /WPE/TestWebKitUserContentManager:/webkit/WebKitUserContentManager/script-message-received, /WPE/TestWebKitWebView:/webkit/WebKitWebView/run-javascript, /WPE/TestWebKitWebView:/webkit/WebKitWebView/disable-web-security, /WPE/TestWebKitWebView:/webkit/WebKitWebView/notification-initial-permission-allowed, /WPE/TestWebKitWebContext:/webkit/WebKitWebContext/uri-scheme, /WPE/TestSSL:/webkit/WebKitWebView/tls-subresource, /WPE/TestDownloads:/webkit/Downloads/policy-decision-download ... (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25634 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19445 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62576 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95727 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19520 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122038 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39692 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29567 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94579 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40075 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97546 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94316 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24092 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39434 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17236 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/35780 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39580 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45068 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39219 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42552 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40958 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/dom/Document.h:
```diff
@@ -2343,7 +2334,7 @@ class Document
     bool m_deferResizeEventForVisibilityChange { false };
 
-    std::unique_ptr<HashMap<String, WeakPtr<Element, WeakPtrImplWithEventTargetData>, ASCIICaseInsensitiveHash>> m_accessKeyCache;
+    std::optional<HashMap<String, WeakPtr<Element, WeakPtrImplWithEventTargetData>, ASCIICaseInsensitiveHash>> m_accessKeyCache;
```
Not sure it matters for Document (I see a distinct lack of packing above indicating size is unlikely to be all that important), but this is slightly bigger than the std::unique_ptr<> since it requires the extra bool. If it doesn't exist already, the HashMap family of types could probably be made Markable<> aware to get this back.

cdumez replied:
I do not think it will matter for Document objects.

darinadler replied:
I did have this exact same thought (including the `Markable` part) while working on this

darinadler committed:
Putting HashMap in a heap block hurts memory use, speed, code size (WebCore DOM)
rdar://155739380
https://bugs.webkit.org/show_bug.cgi?id=295864

Reviewed by Chris Dumez.

HashMap and similar classes are already pointers to a hash table, so there is no reason
to wrap them in a unique_ptr in a heap block and add another level of indirection.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::buildAccessKeyCache): Updated since m_accessKeyCache is
std::optional instead of std::unique_ptr.
(WebCore::Document::invalidateAccessKeyCacheSlowCase): Ditto.
(WebCore::Document::willBeRemovedFromFrame): Updated since m_touchEventTargets
and m_wheelEventTargets are not stored in std::unique_ptr any more.
(WebCore::Document::absoluteRegionForWheelEventTargets): Renamed from
absoluteRegionForEventTargets and updated to use m_wheelEventTargets.

* Source/WebCore/dom/Document.h: Updated m_accessKeyCache to use
std::optional and m_touchEventTargets and m_wheelEventTargets to not
use std::unique_ptr. Deleted unused touchEventTargets and wheelEventTargets
functions. Made absoluteEventRegionForNode private and renamed
absoluteRegionForEventTargets to absoluteRegionForWheelEventTargets
since that's the only thing it's used for.

* Source/WebCore/page/DebugPageOverlays.cpp:
(WebCore::MouseWheelRegionOverlay::updateRegion): Use
absoluteRegionForWheelEventTargets.
* Source/WebCore/page/scrolling/ScrollingCoordinator.cpp:
(WebCore::ScrollingCoordinator::absoluteEventTrackingRegionsForFrame const):
Ditto.

Canonical link: https://commits.webkit.org/297735@main
